File: tcfcli/cmds/native/common/runtime/nodejs6.10/runtime.js
/*
 * by zalezhang 2018/12/17
 */

var fs = require('fs');
var crypto = require('crypto');


var GLOBAL_FUNCTION_HANDLER = process.argv[2] || process.env.SCF_FUNCTION_HANDLER || 'index:main_handler';
var GLOBAL_FUNCTION_NAME = process.env.SCF_FUNCTION_NAME || "main_handler";
var GLOBAL_EVENT_BODY = process.argv[3] || process.env.SCF_EVENT_BODY ||
    (process.env.DOCKER_USE_STDIN && fs.readFileSync('/dev/stdin', 'utf8')) || '{}';


var GLOBAL_USER_FILE_PATH = '';
var GLOBAL_VERSION = process.env.SCF_FUNCTION_VERSION || '$LATEST';
var GLOBAL_MEM_SIZE = process.env.SCF_FUNCTION_MEMORY_SIZE || '256';
var GLOBAL_TIMEOUT = process.env.SCF_FUNCTION_TIMEOUT || '3';
var GLOBAL_ENVIRON = process.env.SCF_FUNCTION_ENVIRON || '';

var GLOBAL_REQUEST_ID =  uuid();
var GLOBAL_START_TIME = process.hrtime();
var GLOBAL_SOCK = -1;
var GLOBAL_STAGE = 0;


module.exports = {
    init: function() {
        consoleLog('START RequestId: ' + GLOBAL_REQUEST_ID);
        return 0;
    }
    ,
    wait_for_invoke: function() {
        var invokeInfo = [];

        GLOBAL_STAGE += 1;
        switch (GLOBAL_STAGE)
        {
            case 1:
                invokeInfo.cmd = 'RELOAD';
                invokeInfo.context = GLOBAL_USER_FILE_PATH + GLOBAL_FUNCTION_HANDLER;
                break;
            case 2:
                invokeInfo.cmd = 'EVENT';
                invokeInfo.sockfd = GLOBAL_SOCK;
                invokeInfo.event = GLOBAL_EVENT_BODY;
                invokeInfo.context = initContext();
                break;
            default:
                process.exit();
        }

        return invokeInfo;
    },
    log: function (str) {
        //consoleLog(formatSystem(str));
    },
    console_log: function (str) {
        consoleLog(formatConsole(str));
    },
    report_fail: function(stackTrace, typeNum, errType) {
        result = {};

        result['errorCode'] = 1;
        result['errorMessage'] = 'user code exception caught';
        if (stackTrace) {
            result['stackTrace'] = stackTrace;
        }

        reportDone("");
        console.dir(result);
    },
    report_running: function () {
        GLOBAL_START_TIME = process.hrtime();
    },
    report_done: function (resultStr, typeNum) {
        reportDone(resultStr);
    },
}

function initContext() {
    var context = {};
    context['function_version'] = GLOBAL_VERSION;
    context['function_name'] = GLOBAL_FUNCTION_NAME;

    context['time_limit_in_ms'] = GLOBAL_TIMEOUT;
    context['memory_limit_in_mb'] = GLOBAL_MEM_SIZE;

    context['request_id'] = GLOBAL_REQUEST_ID;
    context['environ'] = GLOBAL_ENVIRON;

    return JSON.stringify(context)
}

function reportDone(resultStr) {
    var diffMs = hrTimeMs(process.hrtime(GLOBAL_START_TIME));
    var billedMs = Math.min(100 * (Math.floor(diffMs / 100) + 1), GLOBAL_TIMEOUT * 1000);
    consoleLog('END RequestId: ' + GLOBAL_REQUEST_ID);
    consoleLog([
        'REPORT RequestId: ' + GLOBAL_REQUEST_ID,
        'Duration: ' + diffMs.toFixed(2) + ' ms',
        'Billed Duration: ' + billedMs + ' ms',
        'Memory Size: ' + GLOBAL_MEM_SIZE + ' MB',
        'Max Memory Used: ' + Math.round(process.memoryUsage().rss / (1024 * 1024)) + ' MB',
        '',
    ].join('\t'));

    if (typeof resultStr === 'string') {
        consoleLog('\n' + resultStr);
    }
}

function consoleLog(str) {
    process.stdout.write(str + '\n');
}

function formatConsole(str) {
    return str.replace(/^[0-9TZ:.-]+\t[0-9a-f-]+\t/, '\u001b[34m$&\u001b[0m');
}

function formatSystem(str) {
    return '\u001b[32m' + str + '\u001b[0m';
}

function hrTimeMs(hrtime) {
    return (hrtime[0] * 1e9 + hrtime[1]) / 1e6;
}

function uuid() {
    return crypto.randomBytes(4).toString('hex') + '-' +
        crypto.randomBytes(2).toString('hex') + '-' +
        crypto.randomBytes(2).toString('hex').replace(/^./, '1') + '-' +
        crypto.randomBytes(2).toString('hex') + '-' +
        crypto.randomBytes(6).toString('hex');
}

File: tcfcli/cmds/native/common/runtime/python3.6/runtime.py
from __future__ import print_function
import os
import sys
import json
import time
import uuid
import pstool

tcf_stdout = sys.stdout
tcf_stderr = sys.stderr

_GLOBAL_START_TIME = time.time()
_GLOBAL_SOCK = -1
_GLOBAL_STAGE = 0

_GLOBAL_REQUEST_ID = str(uuid.uuid4())
_GLOBAL_FUNCTION_NAME = os.environ.get('SCF_FUNCTION_NAME', 'test')
_GLOBAL_VERSION = os.environ.get('SCF_FUNCTION_VERSION', '$LATEST')
_GLOBAL_MEM_SIZE = os.environ.get('SCF_FUNCTION_MEMORY_SIZE', '256')
_GLOBAL_TIMEOUT = int(os.environ.get('SCF_FUNCTION_TIMEOUT', '3'))
_GLOBAL_IS_QUIET = (os.environ.get('SCF_DISPLAY_IS_QUIET', 'False') == 'True')

_GLOBAL_FUNCTION_HANDLER = sys.argv[1] if len(sys.argv) > 1 else os.environ.get('SCF_FUNCTION_HANDLER',
        'index.main_handler')

_GLOBAL_EVENT_BODY =  sys.argv[2] if len(sys.argv) > 2 else os.environ.get('SCF_EVENT_BODY',
        (sys.stdin.read() if os.environ.get('DOCKER_USE_STDIN', False) else '{}'))


def init():
    global _GLOBAL_IS_QUIET
    if not _GLOBAL_IS_QUIET:
        tcf_print("START RequestId: %s" % _GLOBAL_REQUEST_ID)

    os.environ['SOCKETPATH'] = ''
    os.environ['CONTAINERID'] = ''

    return 0



def init_context():
    context = {}

    context['environ'] = ""
    for env in context['environ'].split(';'):
        if not env:
            continue
        k, v = env.split('=', 2)
        os.environ[k] = v

    context['request_id'] = _GLOBAL_REQUEST_ID

    context['function_version'] = _GLOBAL_VERSION
    context['function_name'] = _GLOBAL_FUNCTION_HANDLER

    context['time_limit_in_ms'] = _GLOBAL_TIMEOUT
    context['memory_limit_in_mb'] = _GLOBAL_MEM_SIZE

    return json.dumps(context)


class InvokeInfo(object):
    pass


def wait_for_invoke():
    global _GLOBAL_STAGE
    _GLOBAL_STAGE += 1

    invoke_info = InvokeInfo()
    if _GLOBAL_STAGE == 1:
        invoke_info.cmd = 'RELOAD'
        invoke_info.sockfd = _GLOBAL_SOCK
        invoke_info.event =_GLOBAL_EVENT_BODY
        invoke_info.context = _GLOBAL_FUNCTION_HANDLER
    elif _GLOBAL_STAGE == 2:
        invoke_info.cmd = 'EVENT'
        invoke_info.sockfd = _GLOBAL_SOCK
        invoke_info.event = _GLOBAL_EVENT_BODY
        invoke_info.context = init_context()
    else:
        os._exit(0)

    return invoke_info


def report_done(msg, err_type=0):
    global _GLOBAL_IS_QUIET
    if _GLOBAL_IS_QUIET:
        if msg:
            if err_type != 0:
                tcf_print_err("%s" % msg)
            else:
                tcf_print("%s" % msg)
            return

    if err_type == 0:
        tcf_print("END RequestId: %s" % _GLOBAL_REQUEST_ID)

    duration = int((time.time() - _GLOBAL_START_TIME) * 1000)
    billed_duration = min(100 * int((duration / 100) + 1), _GLOBAL_TIMEOUT * 1000)
    max_mem = pstool.get_peak_memory() # memory use in MB
    if err_type != 0:
        tcf_print_err(
            "REPORT RequestId: %s Duration: %s ms Billed Duration: %s ms Memory Size: %s MB Max Memory Used: %s MB" % (
                _GLOBAL_REQUEST_ID, duration, billed_duration, _GLOBAL_MEM_SIZE, max_mem
            )
        )
        tcf_print_err("\n")
    else:
        tcf_print(
            "REPORT RequestId: %s Duration: %s ms Billed Duration: %s ms Memory Size: %s MB Max Memory Used: %s MB" % (
                _GLOBAL_REQUEST_ID, duration, billed_duration, _GLOBAL_MEM_SIZE, max_mem
            )
        )
        tcf_print("\n")

    if msg:
        if err_type != 0:
            tcf_print_err("%s" % msg)
        else:
            tcf_print("%s" % msg)


def report_running():
    global _GLOBAL_START_TIME
    _GLOBAL_START_TIME = time.time()


def report_fail(stackTrace, mem_kb, ret_code):
    result = {}
    result['errorCode'] = 1
    result['errorMessage'] = 'user code exception caught'
    if stackTrace:
        result['stackTrace'] = stackTrace

    report_done('', err_type=1)
    tcf_print_err(result)


def console_log(msg, err=False):
    global _GLOBAL_IS_QUIET
    if not _GLOBAL_IS_QUIET:
        if err:
            tcf_print_err(msg)
        else:
            tcf_print(msg)


def log(errMsg):
    pass


def tcf_print(*args, **kwargs):
    print(*args, file=tcf_stdout, **kwargs)
    tcf_stdout.flush()

def tcf_print_err(*args, **kwargs):
    print(*args, file=tcf_stderr, **kwargs)
    tcf_stderr.flush()
